cdap-hbase-compat-0.96/src/main/java/co/cask/cdap/data2/increment/hbase96/IncrementHandler.java

```diff
@@ -120,6 +124,67 @@ public void preGetOp(ObserverContext<RegionCoprocessorEnvironment> ctx, Get get,
   }
 
   @Override
+  public Result preIncrement(ObserverContext<RegionCoprocessorEnvironment> e, Increment increment)
+    throws IOException {
+    // note that in 0.96, there is no hook for increment after the region lock was obtained
+    // therefore there is a possibility of race conditions when using this HBase version.
+
+    boolean isIncrement = increment.getAttribute(HBaseTable.DELTA_WRITE) != null;
+    if (!isIncrement) {
+      return null; // this should only trigger for a readless increment
+    }
+    boolean transactional = state.containsTransactionalFamily(increment.getFamilyCellMap().keySet());
+    byte[] txBytes = increment.getAttribute(TxConstants.TX_OPERATION_ATTRIBUTE_KEY);
+    if (transactional && txBytes == null) {
+      throw new IllegalArgumentException("Attribute " + TxConstants.TX_OPERATION_ATTRIBUTE_KEY
+                                           + " must be set for transactional readless increments");
+    }
+    long tsToAssign;
+    if (transactional) {
+      byte[] wpBytes = increment.getAttribute(HBaseTable.WRITE_POINTER);
+      if (wpBytes == null) {
+        throw new IllegalArgumentException("Attribute " + HBaseTable.WRITE_POINTER
+                                             + " must be set for transactional readless increments");
+      }
+      tsToAssign = Bytes.toLong(wpBytes);
+    } else {
+      tsToAssign = state.getUniqueTimestamp();
+    }
+
+    Get get = new Get(increment.getRow());
```

Comment:
Shall we acquire the row lock to cover the get and put operations? Something like:

```java
region.getRowLock(increment.getRow());
try {
  // wait for all prior MVCC transactions to finish - while we hold the row lock
  // (so that we are guaranteed to see the latest state). Copied from Region server code
  MultiVersionConsistencyControl mvcc = region.getMVCC();
  mvcc.completeMemstoreInsert(mvcc.beginMemstoreInsert());
  
  Get get = new Get(....);
  ...
} finally {
  rowLock.release();
}
```